File: 2020/2/solve_p1.py
#!/usr/bin/env python3

def parse(line):
    bound, char, password = line.split(' ')
    low, high = bound.split('-')
    low, high = int(low), int(high)
    char = char[0]

    return (low, high, char, password)


def is_valid(low, high, char, password):
    count = sum(1 for c in password if c == char)
    return count >= low and count <= high


def main():
    res = open('in.txt', 'r').readlines()
    res = map(parse, res)
    res = map(lambda p: is_valid(*p), res)
    res = sum(1 for x in res if x)

    print(res)


if __name__ == '__main__':
    main()

File: 2020/2/solve_p1.hs
#!/usr/bin/env runhaskell

{-# LANGUAGE OverloadedStrings #-}

import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import Data.Text.Read

data PasswordCheck = Check Int Int Char T.Text
                   deriving (Show)

isValid :: PasswordCheck -> Bool
isValid (Check low high ch t)  = count >= low && count <= high
  where count = (T.length . T.filter (== ch)) t

parse :: T.Text -> PasswordCheck
parse s = Check low high ch password
  where xs = T.words s
        [boundT, charT, password] = xs
        [Right (low, _), Right (high, _)] = map decimal (T.splitOn "-" boundT)
        ch = T.head charT

main :: IO ()
main = do
  entries <- T.lines <$> TIO.readFile "in.txt"

  let res = (length . filter id) $ map (isValid . parse) entries

  print res
